File: framework/tests/ut/passes/src/test_loop_unroll.cpp
/**
 * Copyright (c) 2025 Huawei Technologies Co., Ltd.
 * This program is free software, you can redistribute it and/or modify it under the terms and conditions of
 * CANN Open Software License Agreement Version 2.0 (the "License").
 * Please refer to the License for details. You may not use this file except in compliance with the License.
 * THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND, EITHER EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.
 * See LICENSE in the root of the software repository for the full text of the License.
 */

/*!
 * \file test_loop_unroll.cpp
 * \brief Unit test for LoopUnroll pass.
 */

#include <gtest/gtest.h>
#include "interface/function/function.h"
#include "tilefwk/tilefwk.h"
#include "interface/inner/tilefwk.h"
#include "passes/pass_mgr/pass_manager.h"
#include "interface/configs/config_manager.h"
#include "interface/tensor/symbolic_scalar.h"
#include "interface/interpreter/raw_tensor_data.h"
#define private public
#include "passes/tensor_graph_pass/loop_unroll.h"
#undef private
#include "computational_graph_builder.h"

#include <fstream>
#include <vector>
#include <string>

namespace npu {
namespace tile_fwk {
class LoopUnrollTest : public ::testing::Test {
public:
    static void SetUpTestCase() {}

    static void TearDownTestCase() {}

    void SetUp() override
    {
        Program::GetInstance().Reset();
        config::Reset();
        config::SetHostOption(COMPILE_STAGE, CS_EXECUTE_GRAPH);
        config::SetHostConfig(KEY_STRATEGY, "FunctionUnroll");
        config::SetPlatformConfig(KEY_ENABLE_COST_MODEL, false);
        config::SetHostOption(COMPILE_STAGE, CS_TENSOR_GRAPH);
        std::vector<std::string> funcName = {"TENSOR_main"};
        config::SetPassConfig("FunctionUnroll", "LoopUnroll", "CONVERT_TO_STATIC", funcName);
        int s = 32;
        TileShape::Current().SetVecTile(s, s);
        TileShape::Current().SetCubeTile({s, s}, {s, s}, {s, s});
    }
    void TearDown() override {}
};

// Test nested LOOPs
TEST_F(LoopUnrollTest, TestInnerLoopOrder)
{
    ConfigManager::Instance();
    int vecLen = 128;
    int loopNum = 5;
    int tileNum = 4;
    Tensor inputA(DT_FP32, {loopNum, vecLen}, "inputA");
    Tensor inputB(DT_FP32, {tileNum, vecLen}, "inputB");
    Tensor output(DT_FP32, {1, vecLen}, "out");

    FUNCTION("main", {inputA, inputB}, {output})
    {
        LOOP("Outer", FunctionType::DYNAMIC_LOOP, i, LoopRange(tileNum))
        {
            Tensor tileB(DT_FP32, {1, vecLen}, "tileB");
            LOOP("Inner", FunctionType::DYNAMIC_LOOP, j, LoopRange(1))
            {
                (void)j;
                auto tile = View(inputB, {1, vecLen}, {i, 0});
                tileB = Mul(tile, Element(DataType::DT_FP32, 1.0));
            }

            LOOP("Inner2", FunctionType::DYNAMIC_LOOP, k, LoopRange(loopNum))
            {
                auto tileA = View(inputA, {1, vecLen}, {k, 0});
                tileB = Add(tileA, tileB);
            }

            LOOP("Inner3", FunctionType::DYNAMIC_LOOP, l, LoopRange(1))
            {
                (void)l;
                tileB = Mul(tileB, Element(DataType::DT_FP32, 1.0));
                Assemble(tileB, {i, 0}, output);
            }
        }
    }
}

// Test GetTileShape
TEST_F(LoopUnrollTest, test_only_reshape2)
{
    int b = 1;
    int sq = 128;
    int d = 64;
    int bSq = (b == -1) ? -1 : b * sq;
    std::vector<int64_t> qShape = {b, sq, d};

    Tensor q(DT_FP32, qShape, "q");
    Tensor out(DT_FP32, {bSq, d}, "out");

    FUNCTION("main", {q}, {out})
    {
        Tensor qReshape(DT_FP32, {bSq, d}, "qReshape");
        LOOP("LOOP_RESHAPE", FunctionType::DYNAMIC_LOOP, batchId, LoopRange(0, 1, 1), {}, true)
        {
            (void)batchId;
            qReshape = Reshape(q, {bSq, d}, true);
        }

        LOOP("L0_AF", FunctionType::DYNAMIC_LOOP, batchId, LoopRange(GetInputShape(q, 0)), {}, true)
        {
            Tensor q0 = View(qReshape, {sq, d}, {batchId * sq, 0});
            auto tmp = Exp(q0);
            Assemble(tmp, {batchId * sq, 0}, out);
        }
    }
}

TEST_F(LoopUnrollTest, IsNoOverlapWAWAssembleAttrNull)
{
    ComputationalGraphBuilder G;
    G.AddTensor(DataType::DT_FP32, {8, 8}, "a");
    G.AddTensor(DataType::DT_FP32, {8, 8}, "b");
    G.AddTensor(DataType::DT_FP32, {8, 8}, "c");
    G.AddOp(Opcode::OP_ASSEMBLE, {"a"}, {"b"}, "assemble1");
    G.AddOp(Opcode::OP_ASSEMBLE, {"b"}, {"c"}, "assemble2");
    G.SetInCast({"a"});
    G.SetOutCast({"c"});
    LoopUnroll pass;
    pass.lastWriteMap_[0] = std::make_pair(G.GetTensor("b"), true);
    std::unordered_map<Operation*, std::vector<int64_t>> opDynOffsetMap;
    EXPECT_FALSE(pass.IsNoOverlapWAW(0, G.GetTensor("c"), opDynOffsetMap));
}

// Test IF and nested FUNCTION
TEST_F(LoopUnrollTest, TestLoopIfWithRank)
{
    int s = 32;
    int n = 10;

    Tensor t0(DT_FP32, {n * s, s}, "t0");
    Tensor r0(DT_FP32, {s, s}, "r0");
    Tensor out(DT_FP32, {s, s}, "out");

    ProgramData::GetInstance().AppendInputs({
        RawTensorData::CreateConstantTensor<float>(t0, 1.0),
        RawTensorData::CreateConstantTensor<float>(r0, 0.0),
    });
    ProgramData::GetInstance().AppendOutputs({
        RawTensorData::CreateConstantTensor<float>(out, 0.0),
    });

    int len = 8;
    int three = 3;
    int six = 6;

    FUNCTION("main", {t0, r0}, {out})
    {
        LOOP("L0", FunctionType::DYNAMIC_LOOP, i, LoopRange(len))
        {
            IF(i < three)
            {
                IF(IsLoopEnd(i, len)) { r0 = Add(r0, Element(DataType::DT_FP32, 1.0)); }
                ELSE { r0 = Add(r0, Element(DataType::DT_FP32, 1.0)); }
            }
            ELSE
            {
                IF(i < six) { r0 = Add(r0, Element(DataType::DT_FP32, 0.0)); }
                ELSE
                {
                    Tensor t0v = View(t0, {s, s}, {s * i, 0});
                    r0 = Add(t0v, r0);
                }
            }
        }
        config::SetBuildStatic(true);
        FUNCTION("S1") { out = Add(r0, Element(DataType::DT_FP32, 1.0)); }
    }
}
} // namespace tile_fwk
} // namespace npu

File: python/tests/st/test_loop_unroll.py
#!/usr/bin/env python3
# coding: utf-8
# Copyright (c) 2026 Huawei Technologies Co., Ltd.
# This program is free software, you can redistribute it and/or modify it under the terms and conditions of
# CANN Open Software License Agreement Version 2.0 (the "License").
# Please refer to the License for details. You may not use this file except in compliance with the License.
# THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND, EITHER EXPRESS OR IMPLIED,
# INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.
# See LICENSE in the root of the software repository for the full text of the License.
# -----------------------------------------------------------------------------------------------------------
"""
Add Constant Operator

This file implements and tests the add_constant operator which computes y = x + constant
with dynamic shape and nested loop_unroll.
"""

import os
import pypto
import torch
import numpy as np
from numpy.testing import assert_allclose


@pypto.frontend.jit()
def add_constant_kernel(
    x: pypto.Tensor([pypto.DYNAMIC, pypto.DYNAMIC], pypto.DT_FP32),
    y: pypto.Tensor([pypto.DYNAMIC, pypto.DYNAMIC], pypto.DT_FP32)
):
    """
    Add constant operator: y = x + constant

    Args:
        x: Input tensor with dynamic shape [M, N]
        y: Output tensor with dynamic shape [M, N]
        constant: Constant value to add
    """
    m_dyn = x.shape[0]
    n_dyn = x.shape[1]
    constant = 0.5

    tile_m = 16
    tile_n = 16

    m_loop = (m_dyn + tile_m - 1) // tile_m

    for m_idx, m_unroll_factor in pypto.loop_unroll(m_loop, unroll_list=[4, 2, 1]):
        m_offset = m_idx * tile_m
        m_offset_end = pypto.min(m_offset + tile_m * m_unroll_factor, m_dyn)
        valid_m = m_offset_end - m_offset

        n_loop = (n_dyn + tile_n - 1) // tile_n

        for n_idx, n_unroll_factor in pypto.loop_unroll(n_loop, unroll_list=[4, 2, 1]):
            n_offset = n_idx * tile_n
            n_offset_end = pypto.min(n_offset + tile_n * n_unroll_factor, n_dyn)
            valid_n = n_offset_end - n_offset

            actual_tile_m = tile_m * m_unroll_factor
            actual_tile_n = tile_n * n_unroll_factor

            x_view = pypto.view(
                x, [actual_tile_m, actual_tile_n], [m_offset, n_offset],
                valid_shape=[valid_m, valid_n]
            )

            pypto.set_vec_tile_shapes(actual_tile_m, actual_tile_n)
            result = pypto.add(x_view, constant)

            pypto.assemble(result, [m_offset, n_offset], y)


if __name__ == "__main__":
    device_id = int(os.environ.get('TILE_FWK_DEVICE_ID', 0))
    device = f'npu:{device_id}'
    torch.npu.set_device(device_id)


    m, n = 128, 128
    constant = 0.5

    x = torch.ones(m, n, dtype=torch.float32)
    y = torch.zeros(m, n, dtype=torch.float32)

    x_npu = x.npu()
    y_npu = y.npu()


    add_constant_kernel(x_npu, y_npu)
    torch.npu.synchronize()

    y = y_npu.cpu()

    golden = x.cpu() + constant

    assert_allclose(
        np.array(y),
        np.array(golden.cpu()),
        rtol=1e-3,
        atol=1e-3
    )
